File: src/SSVMtoolbox/matlab/ssvm_predict.m
function [PredictedLabel, ErrRate]=ssvm_predict(label, inst, model)
%==========================================================================
% SSVM Predict
%--------------------------------------------------------------------------
% command example
% [PredictedLabel, ErrRate]=ssvm_predict(TLabel, TInst, model);
%--------------------------------------------------------------------------
% Inputs:
% label           [m x 1] : testing data label
% inst            [m x n] : testing data inputs
% model           [struct]: learning model
%   .w            [n x 1] : normal vector of separating (or response) hyperplane
%   .b            [1 x 1] : bias term
%   .RS           [? x n] : reduced set
%   .params       [struct]: learning parameters
%     .s          [1 x1]  : learning algorithm. 0-SSVM, 1-SSVR (default: 0)
%     .t          [1 x 1] : kernel type. 0-linear, 1-polynomial, 2-radial basis 
%     .c          [1 x 1] : the weight parameter C of SVM
%     .g          [1 x 1] : gamma in kernel function 
%     .r          [1 x 1] : ratio of random subset size to the full data size   
%     .v          [1 x 1] : number of cross-validation folds                    
%     .e          [1 x 1] : epsilon-insensitive value in epsilon-SVR            
%     .z          [1 x 1] : number of slices (only using in regression)        
%     .p          [1 x 1] : 0-turn off pca preprocessing in KSIR               
%                           >0-turn on pca preprocessing in KSIR
%--------------------------------------------------------------------------
% Outputs:
% PredictedLabel  [m x 1] : predicted label
% ErrRate         [1 x 1] : error rate (for classification) or relative 2-norm error (for regression)
%==========================================================================

K = build_ker(model.params, inst, model.RS);

if (model.params.s==0 || model.params.s==1) 
    [ErrRate PredictedLabel]=svmpredict(label, K, model);
elseif (model.params.s==2 || model.params.s==3)
    [ErrRate PredictedLabel]=svrpredict(label, K, model);
end

%==========================================================================
% SSVM and KSIRSVM
%==========================================================================
function [ErrRate PredictedLabel]=svmpredict(label, K, model)

if (model.params.s==1)
    K = K*model.edrs;
    %%%%%%%%%%%%%%%%%%%
    %K = scale(K,0,model.scale);
    %%%%%%%%%%%%%%%%%%%
end
[ErrRate PredictedLabel] = MultiPredictor(K, label, model);
if(isempty(label))                
    ErrRate=NaN;
end


%==========================================================================
% SSVR  and KSIRSVR
%==========================================================================
function [ErrRate PredictedLabel]=svrpredict(label, K, model)

if (model.params.s==3)
    K = K*model.edrs;
    %%%%%%%%%%%%%%%%%%%
    %K = scale(K,0,model.scale);
    %%%%%%%%%%%%%%%%%%%
end
% Prediction
PredictedLabel = K*model.w + model.b;
% calculate ErrRate¡]only if "label" exists¡^
if(~isempty(label))        
    ErrRate=zeros(3, 1);
    % 1-R^2
     ErrRate(1)=mean((label-PredictedLabel).^2)/mean((label-mean(label)).^2);
    % relative 2-norm error
    ErrRate(2) = norm(PredictedLabel - label)/norm(label);
    % mean absoute error
    ErrRate(3) = mean(abs(PredictedLabel - label));
else
    ErrRate=NaN;
end    




%==========================================================================
function [ErrRate Predict_label] = MultiPredictor(TestK, Testlabel, model)

k =length(model.Class);
CI = model.RsEndIndex;
num = length(TestK(:,1));
Vote = zeros(num,k);
count = 1;
for i = 1 : k-1
    for j = i+1 : k
        if ( model.params.s==0 & strcmp(model.Space,'dual') & model.params.t~=2)
            K_col = [CI(i)+1:CI(i+1) CI(j)+1:CI(j+1)];
        else
            K_col = [1:length(TestK(1,:))];
        end
        [Vote] = vote(Vote, TestK(:,K_col), model.w{count}, model.b(count), i, j);
        count = count+1;
    end
end
[Non, Predict_label] = max(Vote');
Predict_label = model.Class(Predict_label);
ErrRate = 1-(length(find(Predict_label==Testlabel)))/num;


%==========================================================================
function [Vote] = vote(Vote, K, w, b, i, j)

Num = length(K(:,1));
Temp_i = find((K*w-b)>=0);
Vote(Temp_i,i) = Vote(Temp_i,i)+1;
Temp_j = setdiff([1:Num],Temp_i);
Vote(Temp_j,j) = Vote(Temp_j,j)+1;

%==========================================================================
function [K, flag] = build_ker(params, u, v)
%  params  [struct]: Learning parameters 
%
%  u,v   - kernel data,                                            
%           u is a [m x n] real number matrix,                      
%           v is a [p x n] real number matrix
%  p     - kernel arguments(it dependents on your kernel type)

flag = 'dual';

if (params.t==1)
    p = params.g;
    K = KGaussian(p, u, v);
elseif (params.t==0)
    [m, n] = size(u);
    if ((n > m) || (params.r < 1))
        K = u*v';
        %disp(['Solving linear SSVM/SSVR in dual space'])
    else
        K = u;
        flag = 'primal';
    end
else
    K = u;
    disp(['Using a specified kernel!'])
end

File: src/SSVMtoolbox/r/ssvm_predict.R
build_ker_pre = function(model, u){
  #  params  List: Learning parameters 
  #
  #  u,v   - kernel data,                                            
  #           u is a [m x n] real number matrix,                      
  #           v is a [p x n] real number matrix
  #  p     - kernel arguments(it depends on your kernel type)
  if(model[["params"]][["t"]]==1){
    p = model[["params"]][["g"]]
    K = KGaussian(p, u, model[["RS"]])
  }else if(model[["params"]][["t"]]==0){
    if(model[["Space"]]=="dual"){
      K = u%*%t(v)
    }else{
      K=u
    }
  }else{
    K = u
    cat("Using a specified kernel!")
  }
  return(K)
}


vote_pre = function(Vote, K, w, b, i, j){
  Num = dim(K)[1]
  Temp_i = which((K%*%w-b) >= 0)
  Vote[Temp_i, i] = Vote[Temp_i, i]+1
  Temp_j = sort(setdiff(1:Num, Temp_i))
  Vote[Temp_j, j] = Vote[Temp_j, j]+1
  return(Vote)
  
}


MultiPredictor_pre = function(TestK, Testlabel, model){
  k = length(model[["Class"]])
  CI = model[["RsEndIndex"]]
  num = length(TestK[,1])
  Vote = matrix(0, num, k)
  count = 1
  for(i in 1 : (k-1)){
    for(j in (i+1) : k){
      if(model[["params"]][["s"]]==0 && model[["Space"]]=="dual" && model[["params"]][["t"]]!=2){
        K_Col = cbind((CI[i]+1):(CI[i+1]), (CI[j]+1):(CI[j+1]))
      }else{
        K_Col = 1:length(TestK[1,])
      }
      Vote = vote_pre(Vote, TestK[, K_Col], model[["w"]][[count]], model[["b"]][count], i, j)
      count = count+1
    }
  }
  Predict_label = apply(Vote, 1, which.max)
  Predict_label = model[["Class"]][Predict_label]
  ErrRate = 1-(length(which(Predict_label==Testlabel)))/num
  r=list(ErrRate, Predict_label)
  return(r)
  
}


svrpredict = function(label, K, model){
  PredictedLabel = K%*%model[["w"]] +model[["b"]]
  if(length(label)>0){
    ErrRate = matrix(0,3,1)
    ErrRate[1] = mean((label-PredictedLabel)^2)/mean((label-mean(label))^2)
    ErrRate[2] = norm(PredictedLabel-label,"2")/norm(label,"2")
    ErrRate[3] = mean(abs(PredictedLabel-label))
  
  }else{
    ErrRate = NaN
  }
  r=list(ErrRate, PredictedLabel)
  return(r)
  
}


svmpredict = function(label, K, model){
  temp =MultiPredictor_pre(K, label, model)
  ErrRate = temp[[1]]
  PredictedLabel = temp[[2]]
  if(length(label) <= 0){
    ErrRate = NaN
  }
  r=list(ErrRate, PredictedLabel)
  return(r)
}


ssvm_predict = function(label, inst, model){
  #==========================================================================
  # SSVM Predict
  #--------------------------------------------------------------------------
  # command example
  # [PredictedLabel, ErrRate]=ssvm_predict(TLabel, TInst, model);
  #--------------------------------------------------------------------------
  # Inputs:
  # label           [m x 1] : testing data label
  # inst            [m x n] : testing data inputs
  # model           [struct]: learning model
  #   .w            [n x 1] : normal vector of separating (or response) hyperplane
  #   .b            [1 x 1] : bias term
  #   .RS           [? x n] : reduced set
  #   .params       [struct]: learning parameters
  #     .s          [1 x1]  : learning algorithm. 0-SSVM, 1-SSVR (default: 0)
  #     .t          [1 x 1] : kernel type. 0-linear, 1-polynomial, 2-radial basis 
  #     .c          [1 x 1] : the weight parameter C of SVM
  #     .g          [1 x 1] : gamma in kernel function 
  #     .r          [1 x 1] : ratio of random subset size to the full data size   
  #     .v          [1 x 1] : number of cross-validation folds                    
  #     .e          [1 x 1] : epsilon-insensitive value in epsilon-SVR            
  #     .z          [1 x 1] : number of slices (only using in regression)        
  #     .p          [1 x 1] : 0-turn off pca preprocessing in KSIR               
  #                           >0-turn on pca preprocessing in KSIR
  #--------------------------------------------------------------------------
  # Outputs:
  # PredictedLabel  [m x 1] : predicted label
  # ErrRate         [1 x 1] : error rate (for classification) or relative 2-norm error (for regression)
  #==========================================================================
  options(warn=-1)
  K = build_ker_pre(model, inst)
  if(model[["params"]][["s"]]==0){
    temp = svmpredict(label, K, model)
    ErrRate = temp[[1]]
    PredictedLabel = temp[[2]]
  }else if(model[["params"]][["s"]]==1){
    temp = svrpredict(label, K, model)
    ErrRate = temp[[1]]
    PredictedLabel = temp[[2]]    
  }
  result = list()
  result[["ErrRate"]] = ErrRate
  result[["PredictedLabel"]] = PredictedLabel
  return(result)
}